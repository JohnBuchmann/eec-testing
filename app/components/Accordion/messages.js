/*
 * Accordion Messages
 * This contains all the text for the Accordion component.
 */
import { defineMessages } from 'react-intl';

export const scope = 'app.components.CustomizedAccordion';

export default defineMessages({
  view: {
    id: `${scope}.view`,
    defaultMessage: 'View',
  },
  hide: {
    id: `${scope}.hide`,
    defaultMessage: 'Hide',
  },
});
