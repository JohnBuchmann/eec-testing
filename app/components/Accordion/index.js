/**
 *
 * Accordion.js
 *
 * A common accordion, send title, icon, clases to use and body content to display
 */

import {
  Accordion,
  AccordionDetails,
  AccordionSummary,
  Box,
  Typography,
} from '@material-ui/core';
import { withStyles } from '@material-ui/core/styles';
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { FormattedMessage } from 'react-intl';
import { getValueIfExists } from 'Utils/propertyValidation';
import messages from './messages';

/**
 * AccordionComponent
 * @param {string} title Title to show
 * @param {string} icon Icon path to display
 * @param {node} accordionBodyContent content React Node view to display on Accordion Deatils (Body)
 * @param {boolean} isAccordionOpen Flag to display Accordion Open or Collapse
 * @param {object{container,icon,titleContainer}} Classes apply to the container
 */
class AccordionComponent extends Component {
  constructor() {
    super();

    this.state = {
      isAccordionOpen: false,
    };
  }

  componentDidMount() {
    this.setState({ isAccordionOpen: this.props.isAccordionOpen });
  }

  handleToggleAccordion = () => {
    const isAccordionOpen = getValueIfExists(
      () => this.state.isAccordionOpen,
      false
    );
    this.setState({
      isAccordionOpen: !isAccordionOpen,
    });
  };

  render() {
    const isAccordionOpen = getValueIfExists(
      () => this.state.isAccordionOpen,
      false
    );
    const toggleText = isAccordionOpen ? messages.hide : messages.view;
    const { classes, title, icon } = this.props;

    return (
      <Accordion square expanded={isAccordionOpen}>
        <AccordionSummary
          id="panel1a-header"
          onClick={this.handleToggleAccordion}
        >
          <Box className={classes.container}>
            <Box className={classes.titleContainer}>
              {icon && <img src={icon} className={classes.icon} alt={icon} />}
              <Typography variant="caption">{title}</Typography>
            </Box>
            <Typography color="primary" variant="caption">
              <FormattedMessage {...toggleText} />
            </Typography>
          </Box>
        </AccordionSummary>
        <AccordionDetails>{this.props.accordionBodyContent}</AccordionDetails>
      </Accordion>
    );
  }
}

AccordionComponent.propTypes = {
  title: PropTypes.string,
  icon: PropTypes.string,
  accordionBodyContent: PropTypes.node,
  isAccordionOpen: PropTypes.bool,
  classes: PropTypes.object,
};

const useStyles = () => ({
  container: {
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'space-between',
    width: '100%',
    alignItems: 'center',
  },
  titleContainer: {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
  },
  icon: {
    width: '1.5rem',
    height: '1.5rem',
    marginRight: '0.75rem',
  },
});

export default withStyles(useStyles)(AccordionComponent);
