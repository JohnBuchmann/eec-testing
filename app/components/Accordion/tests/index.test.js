/**
 * Testing our Accordion component
 */

import React from 'react';
import { fireEvent, render } from 'react-testing-library';
import ShallowRenderer from 'react-test-renderer/shallow';
import { IntlProvider } from 'react-intl';

import Accordion from '../index';

const renderer = new ShallowRenderer();

const title = 'Accordion Test Title';
const accordionBodyContent = <h1>Test</h1>;
const icon = '#';
const isAccordionOpen = false;
const renderComponent = (props = {}) =>
  render(
    <IntlProvider locale="en">
      <Accordion
        title={title}
        icon={icon}
        isAccordionOpen={isAccordionOpen}
        {...props}
      >
        {accordionBodyContent}
      </Accordion>
    </IntlProvider>
  );

describe('<Accordion />', () => {
  describe('#constructor', () => {
    it('should render a <Accordion> tag to change route if the handleRoute prop is specified', () => {
      const { container } = renderComponent({ title });
      expect(container.querySelector('accordion')).toBeDefined();
    });

    it('should render and match the snapshot', () => {
      renderer.render(<Accordion />);
      const renderedOutput = renderer.getRenderOutput();
      expect(renderedOutput).toMatchSnapshot();
    });
  });

  describe('#render', () => {
    it('should fire accordion click event and open it', () => {
      const onClickSpy = jest.fn();
      const { getByRole } = renderComponent({
        onClick: onClickSpy,
      });

      fireEvent.mouseDown(getByRole('button'));
      const buttonSpan = getByRole('button').querySelector('span');
      expect(buttonSpan).toBeDefined();
      buttonSpan.click();
    });
  });
});
