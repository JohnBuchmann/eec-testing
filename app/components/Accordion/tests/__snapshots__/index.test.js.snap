// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Accordion /> #constructor should render and match the snapshot 1`] = `
<AccordionComponent
  classes={
    Object {
      "container": "AccordionComponent-container-6",
      "icon": "AccordionComponent-icon-8",
      "titleContainer": "AccordionComponent-titleContainer-7",
    }
  }
/>
`;
